kv: improve `TestTxnCoordSenderRetries`

nvb opened the pull request:
This PR improves `TestTxnCoordSenderRetries` in two ways, in preparation for its use to be expanded to demonstrate the difference in behavior between isolation levels.

### use local TxnMetrics in TestTxnCoordSenderRetries

The first commit updates `TestTxnCoordSenderRetries` to use a local `kv.DB` with a local `TxnMetrics` for its test transaction. This allows the test to precisely assert on the metrics without having to worry about other transactions in the system affecting them.

### add preemptive refreshes and 1PCs to TestTxnCoordSenderRetries

The second commit adds more testing conditions to TestTxnCoordSenderRetries. Each test now asserts it expectations for preemptive refreshes and for 1-phase commits.

nvb committed:
kv: use local TxnMetrics in TestTxnCoordSenderRetries

This commit updates TestTxnCoordSenderRetries to use a local kv.DB with
a local TxnMetrics for its test transaction. This allows the test to
precisely assert on the metrics without having to worry about other
transactions in the system affecting them.

Epic: None
Release note: None

nvb committed:
kv: add preemptive refreshes and 1PCs to TestTxnCoordSenderRetries

This commit adds more testing conditions to TestTxnCoordSenderRetries.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/102086)
<!-- Reviewable:end -->

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102086#-:-NU7gQOW7bNgM8qOOCsI:b593w57)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender_server_test.go` line 2438 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/102086#-NU4e7H-9tiAJ0jCIiCB:-NU4e7H-9tiAJ0jCIiCC:b-5p15ec) ([raw file](https://github.com/cockroachdb/cockroach/blob/98f97cf15ab51a119405875c0853a227ea8fc7b5/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2438)):*
> ```Go
> 		},
> 		{
> 			name: "write too old with get in the clear",
> ```

Unrelated to your patch, how is this case different from the `priorReads=true` case?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_server_test.go` line 2800 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/102086#-NTo_Mxg-0kqO8L5Rnp5:-NTo_Mxh20idMWkD0nha:b-1mg0l2) ([raw file](https://github.com/cockroachdb/cockroach/blob/98f97cf15ab51a119405875c0853a227ea8fc7b5/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2800)):*
> ```Go
> 				return err
> 			},
> 			expClientRestart: true, // can't refresh
> ```

Do you think it's worth add an expectation for `RefreshFail` and checking it every time there is a client restart? 


<!-- Sent from Reviewable.io -->

nvb committed:
kv: remove redundant test case in TestTxnCoordSenderRetries

This was the same as the "write too old with put after prior read" case.

Epic: None
Release note: None

nvb reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102086#-:-NUF59SVEIyHVU0WxJmT:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender_server_test.go` line 2438 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/102086#-NU4e7H-9tiAJ0jCIiCB:-NUF4Tdt052CV_ylYA38:b-2v76v4) ([raw file](https://github.com/cockroachdb/cockroach/blob/68607244d063cf92e2f878dd3b47e2f489f18104/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2438)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Unrelated to your patch, how is this case different from the `priorReads=true` case?
</blockquote></details>

Good point, it's the same test. I removed it.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_server_test.go` line 2800 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/102086#-NTo_Mxg-0kqO8L5Rnp5:-NUF4tZ_4QlYAQ7-963Y:b-22b486) ([raw file](https://github.com/cockroachdb/cockroach/blob/68607244d063cf92e2f878dd3b47e2f489f18104/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2800)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Do you think it's worth add an expectation for `RefreshFail` and checking it every time there is a client restart? 
</blockquote></details>

That's a good idea. I'll do it in a separate PR though, because I have a sequence of stacked changes to this test that would be impacted by the new assertion.


<!-- Sent from Reviewable.io -->
